Source/DafnyRuntime/DafnyRuntime.go

```diff
@@ -692,183 +692,412 @@ func (seq Seq) VerbatimString(asLiteral bool) string {
  * Arrays
  ******************************************************************************/
 
-// An Array is a Go slice representing a (possibly) multidimensional array,
-// along with metadata.  There aren't any methods for updating; instead, you can
-// update through the pointer returned by Index.
-type Array struct {
-  contents []interface{} // stored as a flat one-dimensional slice
-  dims     []int
-}
-
-func newArray(dims ...Int) *Array {
+// Array is the general interface for arrays. Conceptually, it contains some
+// underlying storage (a slice) of elements, together with a record of the length
+// of each dimension. Thus, this general interface supports 1- and multi-dimensional
+// Dafny arrays.
+//
+// All array indices are given as the Go type "int". This is what the Dafny implementation
+// refers to as "target-language array-index type". However, the dimension lengths
+// are given as "big.Int".
+
+type Array interface {
+  dimensionCount() int
+  dimensionLength(dim int) int
+  ArrayGet1(index int) interface{}
+  ArraySet1(value interface{}, index int)
+  anySlice(lo, hi Int) []interface{}
+  // specializations
+  ArrayGet1Byte(index int) byte
+  ArraySet1Byte(value byte, index int)
+  ArrayGet1Char(index int) Char
+  ArraySet1Char(value Char, index int)
+}
+
+/***** newArray *****/
+
+// Multiply the numbers in "dims" and return the product as an "int".
+// If the produce doesn't fit in an "int", panic with the message that the
+// array-size limit has been exceeded.
+// It is expected that len(dims) is at least 1 and that each number in
+// dims is non-negative.
+func computeTotalArrayLength(dims ...Int) int {
+  product := dims[0]
+  for i := 1; i < len(dims); i++ {
+    product = product.Times(dims[i])
+  }
+  if IntOf(0x8000_0000).Cmp(product) <= 0 {
```

Comment:
Unless you push harder, I feel okay with what's there now. There's a comment above this function that says `dims` is expected to contain non-negative integers. The function is internal to the package and is called only from `NewArrayFromExample`.

I added a comment to `NewArrayFromExample` that states the assumptions and states what is checked. This function is public to the package. We could potentially add run-time checks in it (but note that we're still not checking things like: if both `example` and `init` are non-`nil`, then they are values of the same type).

Reply:
That's just fine. I would really like to harden the API for our runtimes, but I'd also even more love to provide enough functionality so that end users never have to write code that references the runtimes directly at all, so I wouldn't spend effort on it now.